components/builder-scheduler/src/data_store.rs

```diff
@@ -262,6 +265,70 @@ impl DataStore {
+                                   owner_id bigint,
+                                   job_state text,
+                                   project_name text,
+                                   net_error_code int,
+                                   net_error_msg text,
+                                   build_started_at timestamptz,
+                                   build_finished_at timestamptz,
+                                   created_at timestamptz DEFAULT now()
+                            )"#,
+        )?;
+
+        // Insert a new job status
+        migrator.migrate("scheduler",
+                            r#"CREATE OR REPLACE FUNCTION insert_job_status_v1 (
+                               job_id bigint,
+                               owner_id bigint,
+                               job_state text,
+                               project_name text,
+                               net_error_code int,
+                               net_error_msg text,
+                               build_started_at timestamptz,
+                               build_finished_at timestamptz
+                           ) RETURNS SETOF job_status AS $$
+                                   BEGIN
+                                       RETURN QUERY INSERT INTO job_status (job_id, owner_id, job_state, project_name, net_error_code, net_error_msg, build_started_at, build_finished_at)
+                                           VALUES (job_id, owner_id, job_state, project_name, net_error_code, net_error_msg, build_started_at, build_finished_at)
+                                           RETURNING *;
+                                       RETURN;
+                                   END
+                               $$ LANGUAGE plpgsql VOLATILE
+                               "#)?;
+
+        migrator.migrate("scheduler",
+                        r#"CREATE OR REPLACE FUNCTION get_job_status_v1 (integer) RETURNS SETOF job_status AS $$
+                               BEGIN
+                                   RETURN QUERY SELECT * FROM job_status
+                                        ORDER BY created_at ASC
+                                        LIMIT $1;
+                                   RETURN;
+                               END
+                               $$ LANGUAGE plpgsql VOLATILE"#)?;
+
+        migrator.migrate(
+            "scheduler",
+            r#"CREATE OR REPLACE FUNCTION delete_job_status_v1 (integer) RETURNS void AS $$
+                              BEGIN
+                                  DELETE FROM job_status
+                                  WHERE id IN (
+                                      SELECT id
+                                      FROM job_status
+                                      ORDER BY created_at ASC
+                                      LIMIT $1
+                                  );
+                              END
+                              $$ LANGUAGE plpgsql VOLATILE"#,
```

Comment:
Postgres is actually able to inline code and fold function bodies into outer queries in some cases when the function is written in SQL, but can't do that when it's in PL/pgSQL.

Another good thing about writing in SQL is that it forces you to express your problem in terms of set operations, which is almost without exception going to be better than expressing it iteratively. For instance, I helped @raskchanky refactor a PL/pgSQL function that was making n+1 insert statements into an SQL statement that does the same, but in a single operation. True, nothing is stopping you from executing that same single query in a PL/pgSQL function, but that seems like just adding noise at that point.

Certainly, if there are situations where we really do need the capabilities of PL/pgSQL (like exception handling, for example), then by all means, we should use PL/pgSQL. You can do a lot more things in SQL in Postgres these days, though, so that's becoming increasingly unnecessary, particularly for the kinds of things we're doing on this project. That, coupled with the clarity that being forced to express your problem in set-oriented operations, makes me reach for SQL first.